File: src/main/java/com/frightanic/smn/core/GeoAdmin.java
package com.frightanic.smn.core;

import com.frightanic.smn.OpenDataSmnConfiguration;
import com.frightanic.smn.api.SmnData;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

public class GeoAdmin {
  private static final String SMN_DATA_CACHE_KEY = "smn";

  private final Logger logger = LoggerFactory.getLogger(GeoAdmin.class);

  private final DataLoader dataLoader;
  private final LoadingCache<String, SmnData> cache;

  private SmnData lastLoadedSmnData;

  public GeoAdmin(OpenDataSmnConfiguration configuration) {
    dataLoader = new DataLoader(configuration.getDataUrl(), configuration.getDataFileName());
    cache = CacheBuilder.newBuilder().
      expireAfterWrite(configuration.getDataCacheTimeToLiveMinutes(), TimeUnit.MINUTES).
      build(new SmnDataCacheLoader());
  }

  public SmnData getSmnData() throws IOException {
    try {
      return cache.get(SMN_DATA_CACHE_KEY);
    } catch (ExecutionException e) {
      if (lastLoadedSmnData == null) {
        throw new IOException(e.getCause());
      } else {
        logger.warn(
          "Loading data via the cache's data loader failed (cache was empty). Returning last available data" + " set.");
        return lastLoadedSmnData;
      }
    }
  }

  private class SmnDataCacheLoader extends CacheLoader<String, SmnData> {
    public SmnData load(String key) throws IOException {
      SmnData smnData = dataLoader.loadSmnData();
      if (smnData != null) {
        logger.info("Loaded SMN data from geo.admin portal. Keeping an internal copy");
        lastLoadedSmnData = smnData;
      }
      return smnData;
    }
  }
}

File: src/main/kotlin/com/frightanic/smn/core/GeoAdmin.kt
package com.frightanic.smn.core

import com.frightanic.smn.ApplicationConfig
import com.frightanic.smn.api.SmnData
import io.quarkus.cache.Cache
import io.quarkus.cache.CacheName
import jakarta.enterprise.context.ApplicationScoped
import org.slf4j.LoggerFactory
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration


@ApplicationScoped
class GeoAdmin(
    @param:CacheName("smn") private val cache: Cache,
    private val applicationConfig: ApplicationConfig
) {
    private val logger = LoggerFactory.getLogger(GeoAdmin::class.java)
    private var lastLoadedSmnData: SmnData? = null
    private val httpClient: HttpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(10))
        .build()

    fun getSmnData(): SmnData {
        return try {
            cache["smn-data", { loadData() }].await().indefinitely()
        } catch (e: RuntimeException) {
            if (lastLoadedSmnData == null) {
                throw e
            } else {
                logger.warn("Loading data via the cache's data loader failed (cache was empty). " +
                        "Returning last available data set.", e)
                lastLoadedSmnData as SmnData
            }
        }
    }

    private fun loadData(): SmnData {
        logger.info("Loading data from source")
        val request: HttpRequest = HttpRequest.newBuilder()
            .uri(applicationConfig.dataUri())
            .GET()
            .build()
        val smnDataString = httpClient.send(request, HttpResponse.BodyHandlers.ofString()).body()
        logger.info("Successfully fetched {} bytes", smnDataString.length)
        lastLoadedSmnData = SmnData(smnDataString)
        return lastLoadedSmnData as SmnData
    }
}
